services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala

```diff
@@ -0,0 +1,183 @@
+import com.google.cloud.monitoring.v3.{MetricServiceClient, MetricServiceSettings}
+import com.google.monitoring.v3._
+import com.google.protobuf.util.Timestamps
+import com.typesafe.config.Config
+import cromwell.services.instrumentation.InstrumentationService.InstrumentationServiceMessage
+import cromwell.services.instrumentation._
+import cromwell.services.instrumentation.impl.stackdriver.StackdriverConfig._
+import cromwell.services.instrumentation.impl.stackdriver.StackdriverInstrumentationServiceActor._
+import cromwell.util.GracefulShutdownHelper.ShutdownCommand
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+
+
+class StackdriverInstrumentationServiceActor(serviceConfig: Config, globalConfig: Config, serviceRegistryActor: ActorRef) extends Actor {
+  implicit lazy val executionContext = context.dispatcher
+
+  val stackdriverConfig = StackdriverConfig(serviceConfig, globalConfig)
+
+  lazy val projectName: ProjectName = ProjectName.of(stackdriverConfig.googleProject)
+  lazy val credentials = stackdriverConfig.auth.credentials(List(MonitoringScope))
+  lazy val metricLabelsMap = generateMetricLabels()
+
+  var metricsMap = Map.empty[StackdriverMetric, List[Double]]
+
+  // Instantiates a client
+  val metricServiceSettings = MetricServiceSettings.newBuilder.setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build
+  final val metricServiceClient = MetricServiceClient.create(metricServiceSettings)
+
+  // Prepares the monitored resource descriptor
+  lazy val resourceLabels = Map[String, String](("project_id", stackdriverConfig.googleProject))
+  lazy val monitoredResource = MonitoredResource.newBuilder.setType("global").putAllLabels(resourceLabels.asJava).build
+
+  // Give the actor time to warm up, then start sending the metrics to Stackdriver at an interval
+  context.system.scheduler.schedule(InitialDelay, stackdriverConfig.flushRate, self, SendStackdriverMetricCommand)
+
+
+  override def receive = {
+    case SendStackdriverMetricCommand => sendMetricData()
+    case InstrumentationServiceMessage(cromwellMetric) => cromwellMetric match {
+      case CromwellTiming(bucket, value, _) => updateMetricMap(bucket, value.toMillis.toDouble, StackdriverGauge)
+      case CromwellGauge(bucket, value) => updateMetricMap(bucket, value.toDouble, StackdriverGauge)
+      case CromwellCount(bucket, value, _) => updateMetricMap(bucket, value.toDouble, StackdriverCumulative)
+      case CromwellIncrement(bucket) => updateMetricMap(bucket, metricValue = 1D, metricKind = StackdriverCumulative)
+    }
+    case ShutdownCommand =>
+      // flush out metrics (if any) before shut down
+      sendMetricData()
+      metricServiceClient.close()
+      context stop self
+  }
+
+
+  private def labelFromConfig(op: StackdriverConfig => Option[String], key: String): List[(String, String)] = {
+    op(stackdriverConfig).fold(List.empty[(String, String)])(v => List((key.replace("-", "_"), v)))
```

Comment:
It seems weird to me that you're returning a list here, since if you returned an `Option` this could just be:
```scala
def labelFromConfig(op: StackdriverConfig => Option[String], key: String): Option[(String, String)] = 
  op(stackdriverConfig).map(v => List((key.replace("-", "_"), v)))
```

Reply:
Did you mean that I should return `Option[List[(String, String)]]`?

Reply:
Or just `Option[(String, String)]`?

Reply:
Talked F2F, it works with `Option[(String, String)]`